### ISSUE TITLE
JacksonJsonNodeJsonProvider throws class cast exceptions

### ISSUE THREAD

**Issue by ahaverdings-deloitte** [zalando/logbook#1369]
https://github.com/zalando/logbook/blob/bc4afe1b12215f32197449b54ae8f96e6c44be55/logbook-json/src/main/java/org/zalando/logbook/json/JsonPathBodyFilters.java#L92

This JacksonJsonNodeJsonProvider throws class cast exceptions:
class com.fasterxml.jackson.databind.node.ArrayNode cannot be cast to class java.util.List (com.fasterxml.jackson.databind.node.ArrayNode is in unnamed module of loader 'app'; java.util.List is in module java.base of loader 'bootstrap')

Resulting that the body is never transformed when there is one rule applied that handels arrays. Once the exception is throw the original body is returned (and this line also hides the exception pretty well...)
https://github.com/zalando/logbook/blob/786791737dfcf8e83daa7152e381836eeda254cd/logbook-json/src/main/java/org/zalando/logbook/json/JsonPathBodyFilters.java#L112-L113

Issue is that the expected type of (T) would be List<String> but JacksonJsonNodeJsonProvider on createArray() is returning an ArrayNode which is not castable to List.

A solution would be to use JacksonJsonProvider instead of JacksonJsonNodeJsonProvider

**Comment by whiskeysierra** [comment]
Where is this cast to a List happening?

**Comment by ahaverdings-deloitte** [comment]
The cast itself is implicit and part of `JsonPath` (and maybe not much to do about it):
https://github.com/json-path/JsonPath/blob/3f4900d6e7361cd21f0327c828dabf83fd188227/json-path/src/main/java/com/jayway/jsonpath/internal/JsonContext.java#L131
> then:
https://github.com/json-path/JsonPath/blob/3f4900d6e7361cd21f0327c828dabf83fd188227/json-path/src/main/java/com/jayway/jsonpath/JsonPath.java#L225
> then:
https://github.com/json-path/JsonPath/blob/3f4900d6e7361cd21f0327c828dabf83fd188227/json-path/src/main/java/com/jayway/jsonpath/JsonPath.java#L752


the line it starts is here:
https://github.com/zalando/logbook/blob/786791737dfcf8e83daa7152e381836eeda254cd/logbook-json/src/main/java/org/zalando/logbook/json/JsonPathBodyFilters.java#L58

**Comment by whiskeysierra** [comment]
Sounds like a bug in json-path then. Maybe there was a change in Jackson which caused this incompatibility?

**Comment by stefan-g** [comment]
I got the same error. The class cast exception occurs when a node is deleted that does not exists. 

The following test fails:

  @Test
  void test(){
    BodyFilter filter =  JsonPathBodyFilters.jsonPath("$.invalid").delete();
    System.out.println(filter.filter("application/json","{\"key_1\":\"value 1\",\"key_2\":\"value 2\"}"));
  }

I my opinion a node that does not exsists should  be ignored (withoud exception).

**Comment by teggr** [comment]
In the meantime, as a workaround I've used the `JsonBodyFilters` instead of the path:

```
JsonBodyFilters.replaceJsonStringProperty(Set.of("fieldName"), CONTENT_REPLACEMENT_TEXT);
```

**Comment by adamnation** [comment]
Also this log can be supressed by adding this to your logback.xml for now:

```xml
    <!-- quiet down the logging for logbook json filters -->
    <logger name="org.zalando.logbook.json.JsonPathBodyFilters" level="INFO"/>
```

**Comment by xpdemon** [comment]
Root cause identified: this is a json-path bug, not a Logbook one.

`JsonContext.set()/delete()` evaluate the path with `Option.AS_PATH_LIST` and cast the result to `List<String>`. When the path matches nothing and `SUPPRESS_EXCEPTIONS` is set (Logbook always sets it), json-path's `JsonPath.handleMissingPathInContext()` returns `jsonProvider.createArray()` — an `ArrayNode` with the Jackson node providers, which does not implement `java.util.List` — hence the `ClassCastException`. With json-smart the bug is invisible because `JSONArray implements List`.

So with `JsonPathBodyFilters`, every filtered body in which the json-path has no match logs this exception (TRACE level). Since the document mutations happen before the failing path-list collection, matched values are still filtered — the exception only fires when there was nothing to filter.

Fix proposed upstream: json-path/JsonPath#1074 (return an empty `java.util.List` in the `AS_PATH_LIST` branch of `handleMissingPathInContext`).
